app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderViewModel.kt

```diff
@@ -43,101 +49,34 @@ import com.saltedge.authenticator.sdk.tools.keystore.KeyStoreManagerAbs
-                appContext.getString(R.string.connect_status_provider_success).format(connection.name)
-            } else {
-                appContext.getString(R.string.errors_connection_failed)
-            }
-        }
-    override val completeMessage: String
-        get() {
-            return if (viewMode.isCompleteWithSuccess) {
-                appContext.getString(R.string.connect_status_provider_success_description)
-            } else {
-                sessionFailMessage ?: appContext.getString(
-                    R.string.errors_connection_failed_description
-                ) ?: ""
-            }
-        }
-
-    override fun setInitialData(initialConnectData: ConnectAppLinkData?, connectionGuid: GUID?) {
-        if (initialConnectData == null && connectionGuid == null) {
-            viewMode = ViewMode.COMPLETE_ERROR
-        } else if (connectionGuid != null) {
-            this.connection = connectionsRepository.getByGuid(connectionGuid) ?: Connection()
-            viewMode = ViewMode.START_RECONNECT
-        } else {
-            this.initialConnectData = initialConnectData
-            viewMode = ViewMode.START_NEW_CONNECT
-        }
-    }
-
-    override fun getTitleResId(): Int {
-        return if (this.connection.guid.isEmpty()) R.string.connections_new_connection
-        else R.string.actions_reconnect
-    }
-
-    override fun onViewCreated() {
-        when (viewMode) {
-            ViewMode.START_NEW_CONNECT -> performFetchConfigurationRequest()
-            ViewMode.START_RECONNECT -> performCreateConnectionRequest()
-            ViewMode.WEB_ENROLL -> loadWebRedirectUrl()
-            ViewMode.COMPLETE_SUCCESS -> Unit
-            ViewMode.COMPLETE_ERROR -> Unit
-        }
-    }
-
-    override fun onViewClick(viewId: Int) {
-        if (viewId == R.id.mainActionView) viewContract?.closeView()
-    }
 
-    override fun fetchProviderConfigurationDataResult(result: ProviderConfigurationData?, error: ApiErrorData?) {
-        when {
-            error != null -> viewContract?.showErrorAndFinish(error.getErrorMessage(appContext))
-            result != null && result.isValid() -> {
-                result.toConnection()?.let {
-                    this.connection = it
-                    performCreateConnectionRequest()
-                } ?: viewContract?.showErrorAndFinish(appContext.getString(R.string.errors_unable_connect_provider))
-            }
-            else -> viewContract?.showErrorAndFinish(appContext.getString(R.string.errors_unable_connect_provider))
-        }
-    }
-
-    override fun onConnectionCreateFailure(error: ApiErrorData) {
-        viewContract?.showErrorAndFinish(error.getErrorMessage(appContext))
-    }
+    var onCloseEvent = MutableLiveData<ViewModelEvent<Unit>>()
+        private set
+    var showErrorAndFinishEvent = MutableLiveData<ViewModelEvent<String>>()
+        private set
+    var loadUrlInWebViewEvent = MutableLiveData<ViewModelEvent<String>>()
```

Comment:
it is event. don't use imperative. don't indicate where observer should load URL